Enhanced ERA5 data processing: generalize ERA5 extraction and CF conversion to support both ensemble and reanalysis data

divine7022 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
This PR enhances the functionality of `extract.nc.ERA5` and `met2CF.ERA5` to support both ERA5 ensemble and reanalysis datasets. Originally, the functions were tailored only for ensemble data. And refactor the code to fix several issues.
- previously, the `met2CF.ERA5` function made assumptions about the presence and order of ERA5 variables (e.g, `ssrd, tp, d2m, sp`), and a fixed 3-hour timestep. So now made timestep (`1h, 3h, etc.`) to detected dynamically not just 3 hours.
- the function attempted to calculate specific humidity even if required variables (`t2m, d2m, sp`) were missing, causing errors. now only calculated if all required variables are present.
- reanalysis data is typically stored as a 3D , dimensions `[longitude, latitude, time]`. Works correctly with `raster::brick()` which expects this format.
- ensemble data uses a 4D , dimensions `[longitude, latitude, time, ensemble_member]` . `raster::brick()` fails or extracts only a slice unless explicitly adapted. this PR introduces conditional logic to detect and correctly extract the ensemble dimension.
- dynamically extracts origin from units attribute (`hours since...` or `seconds since...`) to construct correct time sequences.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [ ] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I agree that PEcAn Project may distribute my contribution under any or all of
	- the same license as the existing code,
	- and/or the BSD 3-clause license.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

divine7022 committed:
add ens and reanalysis support

divine7022 committed:
update .Rd file

divine7022 committed:
update documentation

divine7022 committed:
update extract.nc.ERA5.Rd

divine7022 committed:
refactor and made fluxiable to process both ens and reanalysis

divine7022 committed:
update met2CF.ERA5.Rd

divine7022 committed:
upadate NEWS.md

divine7022 committed:
update CHANGELOG.md

mdietze commented on modules/data.atmosphere/R/extract_ERA5.R:
```diff
@@ -45,21 +57,41 @@ extract.nc.ERA5 <-
+    )
+    sample_file <- file.path(in.path, paste0(in.prefix, years[1], ".nc"))
+    if (!file.exists(sample_file)) {
+      PEcAn.logger::logger.severe(paste0("Sample file not found: ", sample_file))
```
I find referring to the original met file as the "sample file" to be confusing. If I saw this error message I wouldn't know what it was talking about.

divine7022 replied:
fixed 👍

mdietze commented on modules/data.atmosphere/R/extract_ERA5.R:
```diff
@@ -45,21 +57,41 @@ extract.nc.ERA5 <-
+    ncdf4::nc_close(nc_test)
+    
+    if (is_ensemble) {
+      ensemblesN <- seq(1, 10)
```
might not hurt to detect the ensemble size from the nc file rather than assuming it will always be 10

divine7022 replied:
fixed - now detects actual ensemble size dynamically

mdietze commented on modules/data.atmosphere/R/extract_ERA5.R:
```diff
@@ -85,18 +150,33 @@ extract.nc.ERA5 <-
         foreach::foreach(vname = vars, 
                          .packages=c("Kendall", "ncdf4")) %dopar% {
+                           nc_data <- ncdf4::nc_open(ncfile)
+                           on.exit(ncdf4::nc_close(nc_data), add = TRUE) 
```
will this close the connection at the end of the foreach or at the end of the overall function? If the latter, it would be easy to accidentally open too many connections and throw an error.

divine7022 replied:
yes, the `on.exit()` ensures the netcdf connection is closed when each `%dopar%` iteration completes, regardless of how it exits. each `%dopar%` iteration runs in a _separate environment/process_, so, 
process isolation prevents sharing file handles between workers and automatic cleanup occurs when each worker process completes.

mdietze commented on modules/data.atmosphere/R/extract_ERA5.R:
```diff
@@ -130,22 +210,24 @@ extract.nc.ERA5 <-
+                                       # Calling the met2CF inside extract bc in met process met2CF comes before extract !
+                                       out <- met2CF.ERA5(
+                                         slat[s.ind],
+                                         slon[s.ind],
```
can met2CF.ERA5 take in a vector of locations, because that would be much more efficiently that k copies of the function each trying to separately open the same file in parallel

divine7022 replied:
thank you for the suggestion about vectorizing `met2CF.ERA5` to handle multiple locations. I've analyzed this approach carefully, and I believe we should retain the current parallel-by-site processing for the following reasons,
in current **parallel approach** for (eg, `~125 seconds for 100 sites (8 cores) `) and **vectorized sequential approach** for (eg, `~1000 seconds for 100 sites` ) .so the parallel approach is ~8x faster, which far outweighs any function call overhead.

however, if you still suggest removing the **parallel processing** approach and moving to **sequential processing** by passing entire vectors, please let me know . I'm open to understanding if there are other considerations

infotroph replied:
This might be a case where the fastest answer depends a lot on the conditions of the test, e.g. a small-to-medium file on fast local disk vs. a cluster environment where the a big source file is accessed through a network filesystem or is otherwise disk-bound rather than CPU-bound. 

Not asking you to try to test every possible setup, but the latter example is one where we've seen trouble before and is worth at least thinking through.

dlebauer replied:
This is an interesting discussion and good points, but this level of optimization (vectorized vs. parallel processing) seems outside the scope of this quite valuable PR. To keep scope clear and minimize delay, I’d suggest saving this for future iterations - lerhaps creating a separate feature request labeled as “good first issue” that includes exploring tradeoffs in different environments.

infotroph replied:
Point taken about scope. 

But re labels, a detailed optimization question that requires evaluating performance tradeoffs across multiple use cases and hardware configurations, and therefore also needs an understanding of the the whole workflow to know which cases to prioritize, is a great example of something that is _not_ a good first issue.

mdietze commented on modules/data.atmosphere/R/extract_ERA5.R:
```diff
@@ -130,22 +210,24 @@ extract.nc.ERA5 <-
+                                         slat[s.ind],
+                                         slon[s.ind],
+                                         paste0(year,"-01-01"),
+                                         paste0(year,"-12-31"),
```
Is there a reason we're assuming all years are complete and that we need entire years? This assumption was usually valid when met.process was first written, but these days its pretty common to do short-term nowcasts and forecasts.

divine7022 replied:
fixed - now supports partial years and custom date ranges

divine7022 committed:
add dynamic ens size detection and flexible date range handling

divine7022 committed:
replace hardcoded ens size with dynamic ens_size param

divine7022 committed:
update met2CF.ERA5.Rd

divine7022 committed:
Merge remote-tracking branch 'origin/develop' into refactor/era5-met2cf

divine7022 committed:
add CF unit mapping for 'volume_fraction_of_condensed_water_in_soil'

divine7022 committed:
changed dir naming to identify as Mean for reanlysis data

divine7022 committed:
use stats for namespace clarity

divine7022 committed:
Merge remote-tracking branch 'origin/develop' into refactor/era5-met2cf

dlebauer committed:
Merge branch 'develop' into refactor/era5-met2cf

divine7022 committed:
update roxy

divine7022 committed:
now uses the pecan_standard_met_table for cf maping

divine7022 committed:
add era5 entries

divine7022 committed:
update extract.nc.ERA5.Rd

divine7022 committed:
met2CF.ERA5.Rd

divine7022 committed:
update pecan_standard_met_table.Rd

divine7022 committed:
resolve conflict in NEWS.md

divine7022 committed:
resolve conflict

divine7022 commented:
@dlebauer could you please take a look ? i have added ERA5 entry to `pecan_standard_met_table` and updated `met2CF.ERA5` to use `pecan_standard_met_table`

divine7022 committed:
update roxy

divine7022 committed:
add rlang import to resolve era5 global variable NOTE

divine7022 committed:
udpate extract.nc.ERA5.Rd

divine7022 committed:
call setNames by its namespace

dlebauer reviewed (approved):
Looks great!

infotroph commented on modules/data.atmosphere/R/extract_ERA5.R:
```diff
@@ -45,38 +56,128 @@ extract.nc.ERA5 <-
+      } else {
+        if (verbose) PEcAn.logger::logger.info("processing ERA5 reanalysis data")
+      }
+    }
```
Having a chain of three different conditions that all emit the same message seems pretty confusing to the user

divine7022 replied:
yeah simplified 👍

infotroph commented on modules/data.atmosphere/R/extract_ERA5.R:
```diff
@@ -45,38 +56,128 @@ extract.nc.ERA5 <-
+      } else {
+        # new format: "seconds since YYYY-MM-DD HH:MM:SS" (typically 1970-01-01)
+        timestamp <- as.POSIXct(t, tz = "UTC", origin = tustr[[1]][3])
+      }
```
```suggestion
PEcAn.utils::ud_convert(t, tunits, paste("seconds since", tustr[[1][3]])
```

infotroph replied:
Actually both current and suggested will ignore the HH:MM:SS component of tunits. Maybe `tustr <- strsplit(tunits, " since ")`?

divine7022 replied:
good catch! now splits by `" since "`  and capture `YYYY-MM-DD HH:MM:SS` 👍

infotroph commented on modules/data.atmosphere/R/extract_ERA5.R:
```diff
@@ -85,18 +186,33 @@ extract.nc.ERA5 <-
+                             } else {
+                               brick.tmp <- raster::brick(ncfile, varname = vname)
+                               raster::setZ(brick.tmp, timestamp)
+                             }
```
Does this change the output format compared to the previous version? I can't immediately tell by reading this. If so, this should be loudly advertised as a breaking change -- multiple people have downstream code that would be affected by it.

divine7022 replied:
this is a valid concern but
this is a work around where raster support for 3D(`[lat ,lon, time*ens ]`) and doesn't support of 4D (new ens formate `[lat, lon, valid_time, number]`
Currently fixed the edge case for old ens format `[lat ,lon, time*ens ]` and it work reliably for legacy format too..

infotroph commented on modules/data.atmosphere/R/extract_ERA5.R:
```diff
@@ -85,18 +186,33 @@ extract.nc.ERA5 <-
                              nn[nn == nc_data$var[[vname]]$missval] <- NA
                              # send out the extracted var as a new col
-                             ens.out[[ens]] <- t(nn)
+                             ens.out[[ens]] <- t(nn) 
```
```suggestion
                             ens.out[[ens]] <- t(nn)
```

divine7022 replied:
👍

infotroph commented on modules/data.atmosphere/R/met2cf.ERA5.R:
```diff
@@ -6,97 +6,174 @@
+                       overwrite = FALSE,
+                       verbose = TRUE,
+                       is_ensemble = TRUE,
+                       ens_size = 10) {
```
Note that you could get the same result with fewer parameters by setting ens_size = 1 for the reanalysis case and letting everything else iterate over "each of the 1 ensemble members".

divine7022 replied:
yeah removed! Now use case where
`if (ens_size > 1) { #ens } else { #rean }` in `met2cf.ERA5` but preserves in `extract.nc.ERA5` for robust flexible format detections.

infotroph commented on modules/data.atmosphere/R/met2cf.ERA5.R:
```diff
@@ -6,97 +6,174 @@
+    }
+  }
+  
+  # fallback mappings if table is missing or empty
```
The table is built into the same package, so it seems pretty safe to assume it's available any time this function is. Putting fallbacks here seems more confusing than useful.

divine7022 replied:
removed 👍

infotroph commented on modules/data.atmosphere/R/met2cf.ERA5.R:
```diff
@@ -6,97 +6,174 @@
       error = function(e) {
         PEcAn.logger::logger.severe("Something went wrong during the unit conversion in met2cf ERA5.",
                                     conditionMessage(e))
+        return(NULL)
```
logger.severe always throws an error, so this line won't be reached.
```suggestion
```

divine7022 replied:
removed 👍

infotroph commented on modules/data.atmosphere/R/met2cf.ERA5.R:
```diff
@@ -142,25 +215,20 @@ met2CF.ERA5<- function(lat,
-            seq(0, (length(zoo::index(
-              data.for.this.year.ens
-            )) * 3) - 1 , length.out = length(zoo::index(data.for.this.year.ens))),
+            units = "seconds since 1970-01-01 00:00:00",
```
Why the change from hours since start_date?

divine7022 replied:
The change from `"hours since start_date"` to `"seconds since 1970-01-01"` was made to address fundamental compatibility issues with ERA5 format diversity
so i fixed 3-hour interval assumption `seq(0, (n*3)-1, length.out=n)` assumes exactly 3-hour spacing but ERA5 data can be 1-hourly, 3-hourly, 6-hourly, or daily
```
old ERA5 NetCDF Header:
dimensions: longitude, latitude, time
time:units = "hours since 1900-01-01 00:00:00.0"

new ERA5 NetCDF Header:  
dimensions: longitude, latitude, valid_time [, number]
valid_time:units = "seconds since 1970-01-01"
```

Current approach handles both formats by converting to a standardized Unix timestamp format

`as.numeric(zoo::index(data))` preserves exact timing from processed data and works with any ERA5 format (old/new) and any time interval 
and ECMWF documentation confirms ERA5 time format changed from "hours since 1900-01-01" to "seconds since 1970-01-01", making Unix timestamps the most compatible choice.
https://confluence.ecmwf.int/display/CKB/GRIB+to+netCDF+conversion+on+new+CDS+and+ADS+systems

infotroph reviewed (commented):
Thanks for taking this on. I have some questions about compatibility with previously existing met2cf outputs -- if the format has changed that's potentially a big deal, because lots of folks have stashes of existing CF output that they'd like to keep using and add to. See inline for specifics.

dlebauer commented on modules/data.atmosphere/R/met2cf.ERA5.R:
```diff
@@ -6,97 +6,174 @@
+
+  era5_to_cf <- NULL
+  cf_units_map <- NULL
+  if (exists("pecan_standard_met_table", inherits = TRUE)) {
+    era5_tbl <- pecan_standard_met_table %>%
+      dplyr::filter(!is.na(.data$era5) & nzchar(.data$era5))
+    
+    if (nrow(era5_tbl) > 0) {
+      era5_to_cf <- stats::setNames(era5_tbl$cf_standard_name, era5_tbl$era5)
+      cf_units_map <- stats::setNames(era5_tbl$units, era5_tbl$cf_standard_name)
+    }
+  }
+  
+  # fallback mappings if table is missing or empty
+  if (is.null(era5_to_cf) || length(era5_to_cf) == 0) {
+    era5_to_cf <- c(
+      "t2m" = "air_temperature",
+      "sp" = "air_pressure",
+      "tp" = "precipitation_flux",
+      "u10" = "eastward_wind",
+      "v10" = "northward_wind",
+      "ssrd" = "surface_downwelling_shortwave_flux_in_air",
+      "strd" = "surface_downwelling_longwave_flux_in_air",
+      "swvl1" = "volume_fraction_of_condensed_water_in_soil"
+    )
+  }
+  
+  if (is.null(cf_units_map) || length(cf_units_map) == 0) {
+    cf_units_map <- c(
+      "air_temperature" = "K",
+      "air_pressure" = "Pa",
+      "precipitation_flux" = "kg m-2 s-1",
+      "eastward_wind" = "m s-1",
+      "northward_wind" = "m s-1",
+      "surface_downwelling_shortwave_flux_in_air" = "W m-2",
+      "surface_downwelling_longwave_flux_in_air" = "W m-2",
+      "specific_humidity" = "1",
+      "volume_fraction_of_condensed_water_in_soil" = "1"
+    )
+  }
```
Can this condense to:
```suggestion
era5_tbl <- pecan_standard_met_table %>%
  dplyr::filter(!is.na(.data$era5) & nzchar(.data$era5))

era5_to_cf <- stats::setNames(era5_tbl$cf_standard_name, era5_tbl$era5)
cf_units_map <- stats::setNames(era5_tbl$units, era5_tbl$cf_standard_name)
```

dlebauer reviewed (commented):
Getting closer!

divine7022 committed:
update met2CF.ERA5.Rd

divine7022 committed:
refactor raster logic and removed multiple redundent logs

divine7022 committed:
remove cf mapping fallback and removed is_ensemble param

divine7022 committed:
add check for time dimension before calculating ens size

divine7022 committed:
moved pecan_standard_met_table to seperate pr
